### PR TITLE
[MKL-DNN] Enable s8 support for inner product and 3d input with flatten=false

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
**Major changes:**

- Integrate mkl-dnn signed int8 inner product;
- support 3-d input with flatten=false for both f32/int8 inner product.

This PR can also fix issues in #14407 

@pengzhao-intel @TaoLv @ciyongch 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@reminisce @anirudh2290 @ZhennanQin @ciyongch please help take a review :)

File: src/operator/quantization/quantized_fully_connected.cc [hunk 1]
```diff
@@ -50,11 +50,14 @@ bool QuantizedFullyConnectedShape(const nnvm::NodeAttrs& attrs,
   CHECK(!shape_is_none(in_shape->at(0)))
     << "QuantizedFullyConnectedOp input data shape must be given";
   const mxnet::TShape& dshape = in_shape->at(0);
-  mxnet::TShape wshape = Shape2(param.num_hidden, dshape.ProdShape(1, dshape.ndim()));
-  if (dshape.ndim() != 2) {
-      CHECK(param.flatten)
-        << "QuantizedFullyConnectedOp only supports flatten=true when ishape.ndim()!=2 for now. ";
+  index_t num_input;
+  if (!param.flatten) {
+    num_input = dshape[dshape.ndim()-1];
```
Comment (ZhennanQin) [inline src/operator/quantization/quantized_fully_connected.cc]: Add blank before and after operator.

File: src/operator/quantization/quantized_fully_connected.cc [hunk 2]
```diff
@@ -65,7 +68,13 @@ bool QuantizedFullyConnectedShape(const nnvm::NodeAttrs& attrs,
     SHAPE_ASSIGN_CHECK(*in_shape, i, mxnet::TShape{1});
   }
 
-  SHAPE_ASSIGN_CHECK(*out_shape, 0, mxnet::TShape({dshape[0], wshape[0]}));
+  if (!param.flatten) {
+    TShape result_shape(dshape);
+    result_shape[dshape.ndim()-1] = param.num_hidden;
```
Comment (ZhennanQin) [inline src/operator/quantization/quantized_fully_connected.cc]: Add blank before and after operator.

**Comment by ciyongch** [comment]
LGTM :)

File: src/operator/quantization/quantized_fully_connected.cc (hunk 1 above)
Comment (xinyu-intel) [reply src/operator/quantization/quantized_fully_connected.cc]: added:)

File: src/operator/quantization/quantized_fully_connected.cc [hunk 3]
```diff
@@ -80,9 +89,7 @@ bool QuantizedFullyConnectedType(const nnvm::NodeAttrs& attrs,
   CHECK_EQ(out_type->size(), 3U);
 
 #if MXNET_USE_MKLDNN == 1
-  // TODO(ciyong): currently, only uint8 fully_connected is upported,
-  // int8 fully_connected will be supported after mkldnn v0.18
-  TYPE_ASSIGN_CHECK(*in_type, 0, mshadow::kUint8);
+  CHECK(in_type->at(0) == mshadow::kInt8 || in_type->at(0) == mshadow::kUint8);
```
Comment (TaoLv) [inline src/operator/quantization/quantized_fully_connected.cc]: Add error message here - which type is actually given?

File: src/operator/quantization/quantized_fully_connected.cc [hunk 4]
```diff
@@ -182,7 +189,8 @@ void QuantizedFullyConnectedForwardCPU(const nnvm::NodeAttrs& attrs,
 
   if (dshape.ndim() != 2)
     CHECK(param.flatten)
-      << "QuantizedFullyConnectedOp only supports flatten=true when input_shape!=2 for now. ";
+        << "QuantizedFullyConnectedForwardCPU only supports flatten=true "
+        << "when input_shape!=2 for now. ";
```
Comment (TaoLv) [inline src/operator/quantization/quantized_fully_connected.cc]: `input_shape!=2` seems ambiguous. Use `dshape.ndim() != 2`?

File: src/operator/subgraph/mkldnn/mkldnn_fc.cc [hunk 5]
```diff
@@ -307,9 +302,8 @@ static bool SgMKLDNNFCInferType(const nnvm::NodeAttrs &attrs,
   if (full_param.mkldnn_param.quantized) {
     size_t base_num_inputs = full_param.default_param.no_bias ? 2 : 3;
 
-    // TODO(ciyong): currently, only uint8 fully_connected is upported,
-    // int8 fully_connected will be supported after mkldnn v0.18
-    TYPE_ASSIGN_CHECK(*in_types, 0, mshadow::kUint8);
+    CHECK(in_types->at(0) == mshadow::kInt8 ||
+          in_types->at(0) == mshadow::kUint8);
```
Comment (TaoLv) [inline src/operator/subgraph/mkldnn/mkldnn_fc.cc]: Add descriptive error message.

File: tests/python/quantization/test_quantization.py [hunk 6]
```diff
@@ -377,6 +374,9 @@ def maxabs(a, b):
             assert cond == 0
 
     for qdtype in ['int8', 'uint8']:
+        if is_test_for_mkldnn():
+            check_quantized_fc((32, 512, 2), 100, True, qdtype, flatten=False)
+            check_quantized_fc((32, 512, 2, 2), 100, True, qdtype, flatten=False)
```
Comment (TaoLv) [inline tests/python/quantization/test_quantization.py]: Tests for no_bias=False

**Comment by pengzhao-intel** [comment]
Looks good and all reviewers approved the PR.

Thanks for your contribution. Merging now :)

**Pull request merged**